machine.js

```diff
@@ -0,0 +1,337 @@
+            obj = { tstates: t_state + call_tstates, callback, period: 0 };
+            tstates_callbacks.add(obj);
+
+            return obj;
+        },
+
+        /* Register a callback to be called every call_tstates T-states.
+        * The delay parameter will let us, defer the start of the first call,
+        * without altering the period. This is handy for period signal that changes
+        * values for a short period of time (pulses)
+        */
+        registerTstateInterval: function(callback, call_tstates, delay) {
+            if (call_tstates < 0) {
+                return null;
+            }
+
+            /* If the delay parameter is not defined, set it to 0 */
+            delay = delay || 0;
+
+            const obj = { tstates: t_state + delay + call_tstates, callback, period: call_tstates };
+            tstates_callbacks.add(obj);
+            return obj;
+        },
+
+        removeTstateCallback: function(callback) {
+            if (callback != null) {
+                tstates_callbacks.delete(callback);
+            }
+        },
+
+        interrupt: function(interrupt_vector) {
+            zpu.interrupt(false, interrupt_vector);
+            step_cpu();
+        },
+
+        adjustTStatesWhenHalted: function(end) {
+            const earliest = tstatesutils.getEarliestCallback();
+            if (earliest == null || earliest.tstates > end) {
+                /* No callback or no near callback. Increment the T-states and exit */
+                t_state = end;
+                return;
+            }
+            /* Here, the number of T-state the callback is meant to be executed is in the range
+            * [t_state;end], so it is meant to happen during this iteration.
+            * Jump to that amount and execute instructions following it directly. */
+            t_state = earliest.tstates;
+        }
+    }
+
+    this.tstatesutils = tstatesutils;
+
+    /* Check which scale to use for the video */
+    var scale = 1;
+
+    const mmu = new MMU();
+    const rom = new ROM(this);
+    const ram = new RAM();
+    const pio = new PIO(this);
+    /* Peripherals */
+    const vchip = new VideoChip(this, pio, scale);
+    const uart = new UART(this, pio);
+    const i2c = new I2C(this, pio);
+    const keyboard = new Keyboard(this, pio);
+    const ds1307 = new I2C_DS1307(this, i2c);
+    /* We could pass an initial content to the EEPROM, but set it to null for the moment */
+    const eeprom = new I2C_EEPROM(this, i2c, null);
+    const devices = [ rom, ram, vchip, pio, keyboard, mmu ];
+    const zpu = new Z80({ mem_read, mem_write, io_read, io_write });
+
+    this.mmu = mmu;
+    this.rom = rom;
+    this.ram = ram;
+    this.pio = pio;
+    this.vchip = vchip;
+    this.uart = uart;
+    this.i2c = i2c;
+    this.keyboard = keyboard;
+    this.ds1307 = ds1307;
+    this.eeprom = eeprom;
+    this.devices = devices;
+    this.zpu = zpu;
+
+    var registers = zpu.getState();
+    this.registers = registers;
```

Comment:
Similarly for this, can we make it private and create a public function for retrieving the structure?

Reply:
Fixed.